osgar/drivers/gps.py

```diff
@@ -19,26 +20,64 @@ def checksum(s):
     return b"%02X" % (sum)
 
 
-def str2ms(s):
-    'convert DDMM.MMMMMM string to arc milliseconds(int)'
-    if s == b'':  # unknown position
+def str2deg(s):
+    'convert DDMM.MMMMMM string to deg'
+    assert s != ""
+    assert "." in s, s
+    dm, frac = ('0000' + s).split('.')
+    try:
+        return float(dm[:-2]) + float(dm[-2:] + '.' + frac) / 60
+    except Exception as e:
+        print(e)
         return None
-    dm, frac = (b'0000' + s).split(b'.')
+
+
+def nmea2coord_ms(nmea_data):
+    lon = nmea_data["lon"]
+    lat = nmea_data["lat"]
+    if lon and lat:
+        return [int(round(lon*3_600_000)), int(round(lat*3_600_000))]
+    return [None, None]
+
+
+def parse_nmea(line):
+    nmea_list = line.decode().split(",")
+    assert len(nmea_list) == 15
+    nmea_data = {}
+    # NMEA sentence $GPGGA or $GNGGA
+    # https://docs.novatel.com/OEM7/Content/Logs/GPGGA.htm
     try:
-        return round((int(dm[:-2]) * 60 + float(dm[-2:] + b'.' + frac)) * 60000)
+        nmea_data["identifier"] = nmea_list[0]
+        nmea_data["lon"] = None if nmea_list[4] == "" else str2deg(nmea_list[4])
+        nmea_data["lon_dir"] = None if nmea_list[5] == "" else nmea_list[5]
+        nmea_data["lat"] = None if nmea_list[2] == "" else str2deg(nmea_list[2])
+        nmea_data["lat_dir"] = None if nmea_list[3] == "" else nmea_list[3]
+        nmea_data["utc_time"] = None if nmea_list[1] == "" else datetime.strptime(nmea_list[1], "%H%M%S.%f").time()
+        nmea_data["quality"] = None if nmea_list[6] == "" else int(nmea_list[6])
+        nmea_data["sats"] = None if nmea_list[7] == "" else int(nmea_list[7])
+        nmea_data["hdop"] = None if nmea_list[8] == "" else float(nmea_list[8])
+        nmea_data["alt"] = None if nmea_list[9] == "" else float(nmea_list[9])
+        nmea_data["a-units"] = None if nmea_list[10] == "" else nmea_list[10]
```

Comment:
why do you use minus here as key? (vs. utc_time, lat_dir, etc)

Reply:
It is actually copied from https://docs.novatel.com/OEM7/Content/Logs/GPGGA.htm
I can replace it by "_". No problem.

Reply:
Renamed.